ntoskrnl/config/cmparse.c

```diff
@@ -989,40 +1129,731 @@ CmpHandleExitNode(IN OUT PHHIVE *Hive,
     *Cell = (*Kcb)->KeyCell;
 
-    /* Make sure it's not a dead KCB */
-    ASSERT((*Kcb)->RefCount > 0);
+    /* Do we have any subkeys? */
+    if (!RemainingSubkeysInTotal && !SubkeysInTotal)
+    {
+        /*
+         * We don't have any subkeys nor remaining levels, the
+         * KCB points to the actual key. Lock it.
+         */
+        if (!CmpReferenceKeyControlBlock(*Kcb))
+        {
+            /* This key is opened too many times, bail out */
+            DPRINT1("Could not reference the KCB, too many references (KCB 0x%p)\n", Kcb);
+            return STATUS_UNSUCCESSFUL;
+        }
 
-    /* Reference it */
-    (VOID)CmpReferenceKeyControlBlock(*Kcb);
+        CmpAcquireKcbLockSharedByIndex(GET_HASH_INDEX((*Kcb)->ConvKey));
 
-    /* Return success for now */
-    return STATUS_SUCCESS;
+        /* Add this KCB in the array of locked KCBs */
+        OuterStackArray[0] = 1;
+        OuterStackArray[1] = GET_HASH_INDEX((*Kcb)->ConvKey);
+        *LockedKcbs = OuterStackArray;
+
+        /* And return all the subkey level counters */
+        *TotalRemainingSubkeys = RemainingSubkeysInTotal;
+        *MatchRemainSubkeyLevel = 0;
+        *TotalSubkeys = SubkeysInTotal;
+        return STATUS_SUCCESS;
+    }
+
+    /* Lookup in the cache */
+    Status = CmpLookInCache(HashCacheStack,
+                            FALSE,
+                            RemainingSubkeysInTotal,
+                            Current,
+                            OuterStackArray,
+                            Kcb,
+                            Hive,
+                            Cell,
+                            &MatchRemainingSubkeys);
+    if (!NT_SUCCESS(Status))
+    {
+        /* Bail out if cache lookup failed for other reasons */
+        if (Status != STATUS_RETRY)
+        {
+            DPRINT1("CmpLookInCache() failed (Status 0x%lx)\n", Status);
+            *LockedKcbs = NULL;
+            return Status;
+        }
+
+        /* Retry looking in the cache but with KCBs locked exclusively */
+        Status = CmpLookInCache(HashCacheStack,
+                                TRUE,
+                                RemainingSubkeysInTotal,
+                                Current,
+                                OuterStackArray,
+                                Kcb,
+                                Hive,
+                                Cell,
+                                &MatchRemainingSubkeys);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("CmpLookInCache() failed after retry (Status 0x%lx)\n", Status);
+            *LockedKcbs = NULL;
+            return Status;
+        }
+    }
+
+    /*
+     * Check if we have a full match of remaining levels.
+     *
+     * FIXME: It is possible we can catch a fake key from the cache
+     * when we did the lookup, in such case we should not do any
+     * locking as such KCB does not point to any real information.
+     * Currently ReactOS doesn't create fake KCBs so we are good
```

Comment:
What are those "fake" KCBs about?

Reply:
A key control block has to belong to a registry key that actually must exist in the registry tree, fake keys are the opposite in the spectrum in the sense that they are not real (call them "imaginary" or whatever). We have a very few mentions of this such as `CM_KCB_KEY_NON_EXIST` and `IsFake`constructs (e.g. grep in https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/config/cmkcbncb.c;hb=f8c7bd6898ec15bb8420b7c7a2683759b67ddcd7#l146).

I don't know exactly what's the point of these, perhaps they are simply just dummy KCBs that serve only specific internal kernel purposes. I just discovered that such fake keys are handled in the Windows' registry parser whilst doing some reverse engineering so I thought of adding this FIXME comment as a remainder for whoever wants to work on such thing in the future.